### PR TITLE
feat: add unreachable typematch test

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by mlutze** [comment]
> _ No description provided. _

Doesn't actually catch all unreachable stuff since something like this is not caught:
```scala
typematch x {
    case x: _ => ??? // everything goes here
    case x: List[Int32] => ???
}
```

But the only way I could think of doing that was a n^2 pairwise comparison, so I'm not bothering checking it.

File: main/test/ca/uwaterloo/flix/language/phase/TestRedundancy.scala [hunk 1]
```diff
@@ -1278,4 +1278,51 @@ class TestRedundancy extends FunSuite with TestUtils {
...
+        |
+        |def f(_: a): Unit =
+        |    typematch E {
+        |        case _: E[a] => ()
+        |        case _: _ => ()
+        |    }
+        |""".stripMargin
+    val result = compile(input, Options.TestWithLibNix)
+    expectError[RedundancyError.UnreachableTypeMatchCase](result)
+  }
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestRedundancy.scala:1327]: I was expecting a new more tests given the complexities :)

Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/language/phase/TestRedundancy.scala:1327]: few*

File: main/src/ca/uwaterloo/flix/language/errors/RedundancyError.scala [hunk 2]
```diff
@@ -609,4 +609,30 @@ object RedundancyError {
     def explain(formatter: Formatter): Option[String] = None
   }
 
+  /**
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/RedundancyError.scala:612]: Add over-all comment.

File: main/test/ca/uwaterloo/flix/language/phase/TestRedundancy.scala [hunk 3]
```diff
@@ -1278,4 +1278,103 @@ class TestRedundancy extends FunSuite with TestUtils {
...
+  test("TestUnreachableTypeMatchCase.03") {
+    val input =
+      """
+        |enum E[_] {
+        |    case E
+        |}
+        |
+        |def f(_: a): Unit =
+        |    typematch E {
+        |        case _: E[a] => ()
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/language/phase/TestRedundancy.scala:1321]: What's the reasoning here?

Comment (mlutze) [reply main/test/ca/uwaterloo/flix/language/phase/TestRedundancy.scala:1321]: See above. `a` is bound so it's rigid, acting like the `Int32` in the other example.

**Comment by magnus-madsen** [comment]
I see a test failure.

**Comment by mlutze** [comment]
> I see a test failure.

Yes, this is the problem I don't know how to solve yet :)

**Comment by magnus-madsen** [comment]
> > I see a test failure.
> 
> Yes, this is the problem I don't know how to solve yet :)

Ah

**Comment by magnus-madsen** [comment]
Could some subset of this be merged or?

**Comment by mlutze** [comment]
What I have implemented seems to work for most cases, but I'm not confident in it. I'll close since I think the real solution will likely look different.

**Comment by magnus-madsen** [comment]
OK

**Pull request closed without merging**